File: booleans/booleans.py
########################################################################
#
#  EXERCISE 1 - INITALIZATION
#
#  Initialize the 'newBoolean' variable as a boolean literal of value False
#
###################### YOUR CODE BELOW THIS LINE #######################

newBoolean = False

########################################################################



########################################################################
#
#  EXERCISE 2 - DO I PASS THIS COURSE
#
#  To graduate, the student needs at least one of their exams to be above 75%
#  pass mark. Check to see if the student passes at least one of their exams
#  ('examScore1' and 'examScore2') and save the result into 'studentPasses'
#
import random
examScore1 = random.randint(0, 100)
examScore2 = random.randint(0, 100)
scoreToPass = 75
#
###################### YOUR CODE BELOW THIS LINE #######################

studentPasses = examScore1 >= 75 or examScore2 >= 75

########################################################################



########################################################################
#
#  EXERCISE 3 - HOW MUCH WOOD WOULD A WOODCHUCK CHUCK?
#
#  Two children were asked to guess how many times the characters "wood"
#  and "chuck" (but NOT in the word "woodchuck") appeared in the famous
#  'tongueTwiser'. You have two tasks:
#  - Get the correct number of times "wood" and "chuck" appeared in 'tongueTwiser',
#    not including in "woodchuck", and save them into 'numberOfWood' and 'numberOfChuck' respectively
#  - Confirm they BOTH GOT AT LEAST ONE GUESS CORRECT for either
#    occurence of "wood" or "chuck" and save it into 'bothAtLeastOneCorrect'
#
# tongueTwiser was concatenated like this so the string does not go out of view
tongueTwiser = "How much wood would a woodchuck chuck if a woodchuck could chuck wood? "
tongueTwiser += "He would chuck, he would, as much as he could, and chuck as much wood, "
tongueTwiser += "as a woodchuck would if a woodchuck could chuck wood."
# Student 1's guesses
student1Wood = random.randint(3, 5)
student1Chuck = random.randint(4, 6)
# Student 2's guesses
student2Wood = random.randint(3, 5)
student2Chuck = random.randint(4, 6)
#
###################### YOUR CODE BELOW THIS LINE #######################

numberOfWood = tongueTwiser.count('wood ') + tongueTwiser.count('wood,') + tongueTwiser.count('wood.') + tongueTwiser.count('wood?')
numberOfChuck = tongueTwiser.count(' chuck')

bothAtLeastOneCorrect = False

########################################################################



########################################################################
#
#  EXERCISE 4 - TRUTHY AND FALSY
#
#  As well as a type, each value also has an inherent boolean value,
#  generally known as either truthy or falsy. Some of the rules are a
#  little bizarre so have a look at the following conversions using
#  the bool() conversion function, which will look at the non-boolean
#  data types and return their inherent boolean value
#
falsy1 = falsy2 = falsy3 = None
truthy1 = truthy2 = None
# Uncomment the lines below to see what they evaluate to
#
###################### YOUR CODE BELOW THIS LINE #######################

falsy1 = bool("")
falsy2 = bool(0)
falsy3 = bool(None)

truthy1 = bool("non-empty string")
truthy2 = bool(1)

########################################################################



###################### END OF BOOLEANS EXERCISES #######################



########################################################################
#
#                               ANSWERS
#
#  DO NOT LOOK AT THIS SECTION UNTIL YOU HAVE FINISHED THE EXERCISES
#
#  The following code tests the variables you need to work on to confirm your work is
#  correct. The if conditions check the variables are correct and the print()
#  statements are what make the print outs you see when you run the file. The below
#  is a crude demonstration of what we refer to as code tests. Code testing is a
#  common practice in the software engineer's role, where tests are run to confirm all
#  code runs correctly. Only when all tests pass would we give ourselves the green light
#  to ship code to production and feel confident the system will perform as expected.
#
########################################################################

# setGreen instructs the shell to set the printed output to green
setGreen = "\x1b[32m"
# setRed instructs the shell to set the printed output to red
setRed = "\x1b[31m"
# resetColor instructs the shell to reset the printout color to standard
resetColor = "\x1b[0m"

# Set correctMsg and incorrectMsg Messages so I don't need to write it everytimne
correctMsg = setGreen + "Correct! " + resetColor
incorrectMsg = setRed + "Incorrect! " + resetColor

# totalTests and passedTests count the final tally of passed tests
totalTests = 0
passedTests = 0


# Boolean Initialization
if newBoolean == None:
	print(incorrectMsg + "The initialized 'newBoolean' variable should be defined")
elif type(newBoolean) is bool and not newBoolean:
	print(correctMsg + "The value of 'newBoolean' you got was: '" + str(newBoolean) + "'")
	passedTests+= 1
else:
	print(incorrectMsg + "The value of 'newBoolean' should be 'False', yours is '" + str(newBoolean) + "'")
totalTests += 1


# Do I pass The Course
correctStudentPasses = examScore1 >= scoreToPass or examScore2 >= scoreToPass
if studentPasses == None:
	print(incorrectMsg + "The initialized 'studentPasses' variable should be defined")
elif type(studentPasses) is bool and studentPasses == correctStudentPasses:
	print(correctMsg + "Passed exam score tests")
	print("\tThe students two scores on the test were " + str(examScore1) + " and " + str(examScore2) + ", the score to pass was " + str(scoreToPass) + ".")
	print("\tThe value of 'studentPasses' you got was: '" + str(studentPasses) + "'")
	passedTests+= 1
else:
	print(incorrectMsg + "Failed exam score tests")
	print("\tThe students two scores on the test were " + str(examScore1) + " and " + str(examScore2) + ", the score to pass was " + str(scoreToPass) + ".")
	print("\tThe value of 'studentPasses' should be '" + str(correctStudentPasses) + "', yours is '" + str(studentPasses) + "'")
totalTests += 1


# How Much Wood Would a Woodchuck Chuck
correctNumberOfWood = tongueTwiser.count("wood ") + tongueTwiser.count("wood.") + tongueTwiser.count("wood?") + tongueTwiser.count("wood,")
correctNumberOfChuck = tongueTwiser.count(" chuck")
correctBothAtLeast = (student1Chuck == correctNumberOfChuck or student1Wood == correctNumberOfWood) and (student2Chuck == correctNumberOfChuck or student2Wood == correctNumberOfWood)
if numberOfWood == None or numberOfChuck == None or bothAtLeastOneCorrect == None:
	print(incorrectMsg + "The initialized 'numberOfWood', 'numberOfChuck' and 'bothAtLeastOneCorrect' variables should be defined")
elif numberOfWood == correctNumberOfWood and numberOfChuck == correctNumberOfChuck and type(bothAtLeastOneCorrect) is bool and bothAtLeastOneCorrect == correctBothAtLeast:
	print(correctMsg + "There were exactly " + str(correctNumberOfWood) + " woods and " + str(correctNumberOfChuck) + " chucks in the sentence.")
	print("\tYou said there were " + str(numberOfWood) + " woods and " + str(numberOfChuck) + " chucks in the sentence.")
	print("\tStudent 1 guessed " + str(student1Wood) + " woods and " + str(student1Chuck) + " chucks!")
	print("\tStudent 2 guessed " + str(student2Wood) + " woods and " + str(student2Chuck) + " chucks!")
	print("\tThe value of 'bothAtLeastOneCorrect' you got was thus: '" + str(bothAtLeastOneCorrect) + "' on whether both had at least one answer correct")
	passedTests+= 1
else:
	print(incorrectMsg + "There were exactly " + str(correctNumberOfWood) + " woods and " + str(correctNumberOfChuck) + " chucks in the sentence.")
	print("\tYou said there were " + str(numberOfWood) + " woods and " + str(numberOfChuck) + " chucks in the sentence.")
	print("\tStudent 1 guessed " + str(student1Wood) + " woods and " + str(student1Chuck) + " chucks!")
	print("\tStudent 2 guessed " + str(student2Wood) + " woods and " + str(student2Chuck) + " chucks!")
	print("\tThe value of 'bothAtLeastOneCorrect' should thus be '" + str(correctBothAtLeast) +  "' but you got '" + str(bothAtLeastOneCorrect) + "'")
totalTests += 1


# Truty and Falsy Values
if falsy1 == None or falsy2 == None or falsy3 == None or truthy1 == None or truthy2 == None:
	print(incorrectMsg + "The initialized 'falsy1', 'falsy2', 'falsy3', 'falsy4', 'truthy1', and 'truthy2' variables should be defined")
elif not falsy1 and not falsy2 and not falsy3 and truthy1 and truthy2:
	print(correctMsg + "Passed falsy and truthy values:")
	print("\tFalsy values:")
	print('\tThe expression [ bool("") ]          evaluates to ' + str(falsy1))
	print('\tThe expression [ bool(0) ]           evaluates to ' + str(falsy2))
	print('\tThe expression [ bool(None) ]        evaluates to ' + str(falsy3))
	print("\tTruthy values:")
	print('\tThe expression [ bool("non-empty") ] evaluates to ' + str(truthy1))
	print('\tThe expression [ bool(1) ]           evaluates to ' + str(truthy2))
	passedTests+= 1
else:
	print(incorrectMsg + "Failed falsy and truthy values:")
	print("\tFalsy values:")
	print("\tThe expression [ bool(\"\") ] should be saved into 'falsy1'")
	print("\tThe expression [ bool(0) ] should be saved into 'falsy2'")
	print("\tThe expression [ bool(None) ] should be saved into 'falsy3'")
	print("\tTruthy values:")
	print("\tThe expression [ bool(\"non-empty\") ] should be saved into 'truthy1'")
	print("\tThe expression [ bool(1) ] should be saved into 'truthy2'")
totalTests += 1


# Final Test Tally Determination
if passedTests == totalTests:
	print(setGreen + "You passed " + str(passedTests) + " / " + str(totalTests) + " tests. Congratulations!" + resetColor)
else:
	print(setRed + "You passed " + str(passedTests) + " / " + str(totalTests) + " tests. Keep Going!" + resetColor)

File: booleans/booleans.js
////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 1 - INITALIZATION
//
//  Initialize the 'newBoolean' variable as a boolean literal of value true
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let newBoolean = true

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 2 - MOM VS DAD HEIGHT
//
//  We generate mom and dad's height randomly, use comparison operators to:
//  - Check if 'momsHeight' was greater than 'dadsHeight', saved into 'momIsTaller'
//  - Check if 'dadsHeight' was greater than 'momsHeight', saved into 'dadIsTaller'
//  - Check if 'dadsHeight' equals 'momsHeight', saved into 'areTheSameHeight'
let momsHeight = generateRandomNumber(100, 200)
let dadsHeight = generateRandomNumber(100, 200)
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let momIsTaller = momsHeight > dadsHeight
let dadIsTaller = dadsHeight > momsHeight
let areTheSameHeight = dadsHeight == momsHeight

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 3 - NOT PLAGIARISM?
//
//  As the teacher you are grading student's book reports. You heard through
//  the grapevine students found a website online to copy from, and found a key
//  phrase you think many students would use. TO ensure they did not, check
//  if 'studentEssay' DO NOT contain the substring 'keyPhrase'.
let keyPhrase = "demonstrates his nostalgic tendencies"
let studentEssay = generateSudentEssay()
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let notPlagiarised = !studentEssay.includes(keyPhrase)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 4 - SIBLINGS HEIGHT
//
//  We generate sister's and brother's height randomly, use logical operators to:
//  - Check if BOTH 'sistersHeight' and 'brothersHeight are above the 'thresholdForTall', saved into 'siblingsAreTall'
//  - Check if BOTH 'sistersHeight' and 'brothersHeight are below the 'thresholdForTall', saved into 'siblingsAreShort'
//  - Check if AT LEAST ONE of them are taller than 'thresholdForTall', saved into 'atLeastOneIsTall'
//  - Check if only one of 'sistersHeight' or 'brothersHeight are above the 'thresholdForTall'
//    BUT NOT BOTH, saved into 'onlyOneIsTall'
let sistersHeight = generateRandomNumber(100, 200)
let brothersHeight = generateRandomNumber(100, 200)
let thresholdForTall = 150
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let siblingsAreTall = sistersHeight > thresholdForTall && brothersHeight > thresholdForTall
let siblingsAreShort = sistersHeight < thresholdForTall && brothersHeight < thresholdForTall
let atLeastOneIsTall = sistersHeight > thresholdForTall || brothersHeight > thresholdForTall
let onlyOneIsTall = !siblingsAreTall && !siblingsAreShort

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 5 - VALIDATE A SECRET CODE
//
//  Your friend in class sends you a note with a secret code, you need to
//  validate this 'secretCode', the validation rules are as follows:
//  - The 5th character must be an upper case letter
//  - The 10th character must be a digit
//  - The code must be 15 characters long
//  All of these conditions must be met for the code to be validated, save
//  the result into 'codeIsValid'
let secretCode = generateSecretCode()
//  Hint: This boolean expression is somewhat complicated, you could create
//  numerous variables for each check, and combine them at the end
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let correct5thchar = secretCode.charAt(4) >= 65 && secretCode.charAt(4) <=90
let correct10thchar = secretCode.charAt(11) >= 47 && secretCode.charAt(11) <= 56
let codeIsValid = correct5thchar && correct10thchar

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 6 - TRUTHY AND FALSY
//
//  As well as a type, each value also has an inherent boolean value,
//  generally known as either truthy or falsy. Some of the rules are a
//  little bizarre so have a look at the following conversions using
//  the Boolean() conversion function, which will look at the non-boolean
//  data types and return their inherent boolean value
let falsy1, falsy2, falsy3, falsy4, falsy5
let truthy1, truthy2
//  Uncomment the lines below to see what they evaluate to
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

falsy1 = Boolean("")
 falsy2 = Boolean(0)
 falsy3 = Boolean(undefined)
 falsy4 = Boolean(null)

 truthy1 = Boolean("non-empty string")
 truthy2 = Boolean(1)

////////////////////////////////////////////////////////////////////////



/////////////////////// END OF BOOLEANS EXERCISES //////////////////////



////////////////////////////////////////////////////////////////////////
//
//                               ANSWERS
//
//  DO NOT LOOK AT THIS SECTION UNTIL YOU HAVE FINISHED THE EXERCISES
//
//  The following code tests the variables you need to work on to confirm your work is
//  correct. The if conditions check the variables are correct and the console.log()
//  statements are what make the print outs you see when you run the file. The below
//  is a crude demonstration of what we refer to as code tests. Code testing is a
//  common practice in the software engineer's role, where tests are run to confirm all
//  code runs correctly. Only when all tests pass would we give ourselves the green light
//  to ship code to production and feel confident the system will perform as expected.
//
////////////////////////////////////////////////////////////////////////

// setGreen instructs the shell to set the printed output to green
let setGreen = "\x1b[32m"
// setRed instructs the shell to set the printed output to red
let setRed = "\x1b[31m"
// resetColor instructs the shell to reset the printout color to standard
let resetColor = "\x1b[0m"

// Set correctMsg and incorrectMsg Messages so I don't need to write it everytimne
let correctMsg = setGreen + "Correct! " + resetColor
let incorrectMsg = setRed + "Incorrect! " + resetColor

// totalTests and passedTests count the final tally of passed tests
let totalTests = 0
let passedTests = 0


// String Initialization
if (newBoolean !== undefined && newBoolean) {
    console.log(correctMsg + "The initialized 'newBoolean' is '" + newBoolean + "'")
    passedTests++
} else {
    console.log(incorrectMsg + "The initialized 'newBoolean' should be defined and be 'true', but yours is '" + newBoolean + "'")
}
totalTests++

// Mom vs Dad height
let correctMomIsTaller = momsHeight > dadsHeight
let correctDadIsTaller = dadsHeight > momsHeight
let correctSameHeight = momsHeight === dadsHeight
if (momIsTaller !== undefined && momIsTaller === correctMomIsTaller &&
    dadIsTaller !== undefined && dadIsTaller === correctDadIsTaller &&
    areTheSameHeight !== undefined && areTheSameHeight === correctSameHeight) {
    console.log(correctMsg + "Passed Mom vs data height:")
    console.log("\tMom's height was " + momsHeight + "cm and dad's height was " + dadsHeight + "cm")
    console.log("\tValue of 'momIsTaller' you got was '" + momIsTaller + "'")
    console.log("\tValue of 'dadIsTaller' you got was '" + dadIsTaller + "'")
    console.log("\tValue of 'areTheSameHeight' you got was'" + areTheSameHeight + "'")
    passedTests++
} else {
    console.log(incorrectMsg + "Failed Mom vs data height:")
    console.log("\tMom's height was " + momsHeight + "cm and dad's height was " + dadsHeight + "cm")
    console.log("\t'momIsTaller' should be '" + correctMomIsTaller + "' and you got '" + momIsTaller + "'")
    console.log("\t'dadIsTaller' should be '" + correctDadIsTaller + "' and you got '" + dadIsTaller + "'")
    console.log("\t'areTheSameHeight' should be '" + correctSameHeight + "' and you got '" + areTheSameHeight + "'")
}
totalTests++

// Not plagiarism
let correctNotPlagiarised = !studentEssay.includes(keyPhrase)
if (notPlagiarised !== undefined && notPlagiarised === correctNotPlagiarised) {
    console.log(correctMsg + "The value of 'notPlagiarised' you got was '" + notPlagiarised + "'")
    passedTests++
} else {
    console.log(incorrectMsg + "The value of 'notPlagiarised' should be '" + correctNotPlagiarised + "' but you got '" + notPlagiarised + "'")
}
totalTests++

// Siblings height
let correctAllAreTall = sistersHeight >= thresholdForTall && brothersHeight >= thresholdForTall
let correctAllAreShort = sistersHeight < thresholdForTall && brothersHeight < thresholdForTall
let correctAtLeastOneIsTall = sistersHeight >= thresholdForTall || brothersHeight >= thresholdForTall
let correctOnlyOneIsTall = (sistersHeight > thresholdForTall && brothersHeight < thresholdForTall) || (sistersHeight < thresholdForTall && brothersHeight > thresholdForTall)

if (siblingsAreTall !== undefined && siblingsAreTall === correctAllAreTall &&
    siblingsAreShort !== undefined && siblingsAreShort === correctAllAreShort &&
    atLeastOneIsTall !== undefined && atLeastOneIsTall === correctAtLeastOneIsTall &&
    onlyOneIsTall !== undefined && onlyOneIsTall === correctOnlyOneIsTall) {
    console.log(correctMsg + "Passed siblings height:")
    console.log("\tSister's height was " + sistersHeight + "cm and brother's height was " + brothersHeight + "cm, threshold was " + thresholdForTall + "cm.")
    console.log("\tValue of 'siblingsAreTall' you got was '" + siblingsAreTall + "'")
    console.log("\tValue of 'siblingsAreShort' you got was '" + siblingsAreShort + "'")
    console.log("\tValue of 'atLeastOneIsTall' you got was'" + atLeastOneIsTall + "'")
    console.log("\tValue of 'onlyOneIsTall' you got was'" + onlyOneIsTall + "'")
    passedTests++
} else {
    console.log(incorrectMsg + "Failed siblings height:")
    console.log("\tSister's height was " + sistersHeight + "cm and brother's height was " + brothersHeight + "cm, threshold was " + thresholdForTall + "cm.")
    console.log("\t'siblingsAreTall' should be '" + correctAllAreTall + "' and you got '" + siblingsAreTall + "'")
    console.log("\t'siblingsAreShort' should be '" + correctAllAreShort + "' and you got '" + siblingsAreShort + "'")
    console.log("\t'atLeastOneIsTall' should be '" + correctAtLeastOneIsTall + "' and you got '" + atLeastOneIsTall + "'")
    console.log("\t'onlyOneIsTall' should be '" + correctOnlyOneIsTall + "' and you got '" + onlyOneIsTall + "'")
}
totalTests++

// Validating a secret code
let correctCodeIsValid = (secretCode.charCodeAt(4) >= 65 && secretCode.charCodeAt(4) <= 90) &&
                         (secretCode.charCodeAt(9) >= 48 && secretCode.charCodeAt(9) <= 57) &&
                         secretCode.length === 15
if (codeIsValid !== undefined && codeIsValid === correctCodeIsValid) {
    console.log(correctMsg + "The code was '" + secretCode + "' so value of 'codeIsValid' you got was '" + codeIsValid + "'")
    passedTests++
} else {
    console.log(incorrectMsg + "The code was '" + secretCode + "' so the value of 'codeIsValid' should be '" + correctCodeIsValid + "' but you got '" + codeIsValid + "'")
}
totalTests++

// Truthy vs Falsy
if (falsy1 !== undefined && !falsy1 &&
    falsy2 !== undefined && !falsy2 &&
    falsy3 !== undefined && !falsy3 &&
    falsy4 !== undefined && !falsy4 &&
    truthy1 !== undefined && truthy1 &&
    truthy2 !== undefined && truthy2) {
    console.log(correctMsg + "Passed falsy and truthy values:")
    console.log("\tFalsy values:")
    console.log('\tThe expression [ Boolean("") ]          evaluates to ' + falsy1)
    console.log('\tThe expression [ Boolean(0) ]           evaluates to ' + falsy2)
    console.log('\tThe expression [ Boolean(undefined) ]   evaluates to ' + falsy3)
    console.log('\tThe expression [ Boolean(null) ]        evaluates to ' + falsy4)
    console.log("\tTruthy values:")
    console.log('\tThe expression [ Boolean("non-empty") ] evaluates to ' + truthy1)
    console.log('\tThe expression [ Boolean(1) ]           evaluates to ' + truthy2)
    passedTests++
} else {
    console.log(incorrectMsg + "Failed falsy and truthy values:")
    console.log("\tFalsy values:")
    console.log("\tThe expression [ Boolean(\"\") ] should be saved into 'falsy1'")
    console.log("\tThe expression [ Boolean(0) ] should be saved into 'falsy2'")
    console.log("\tThe expression [ Boolean(undefined) ] should be saved into 'falsy3'")
    console.log("\tThe expression [ Boolean(null) ] should be saved into 'falsy4'")
    console.log("\tTruthy values:")
    console.log("\tThe expression [ Boolean(\"non-empty\") ] should be saved into 'truthy1'")
    console.log("\tThe expression [ Boolean(1) ] should be saved into 'truthy2'")
}
totalTests++

// Final Test Tally Determination
if (passedTests === totalTests) {
    console.log(setGreen + "You passed " + passedTests + " / " + totalTests + " tests. Congratulations!" + resetColor)
} else {
    console.log(setRed + "You passed " + passedTests + " / " + totalTests + " tests. Keep Going!" + resetColor)
}


//////////////////////////// END OF ANSWERS ////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//                           HELPER FUNCTIONS
//
//  FEEL FREE TO LOOK AT THIS SECTION
//
//  The following code generates some variables that you need to check.
//  So you do not cheat, they generate them on a random basis that you
//  cannot predict. So you cannot and must perform the exercises the
//  right way. You can look at them though, they leverage conditionals
//  (if - else) to generate the values used in the tests, and we will
//  see conditionals in the 'Control Flows' chapter of this course
//
////////////////////////////////////////////////////////////////////////


// generateRandomNumber returns a random number between min and max (included)
// so that you cannot cheat and write boolean literals based on the exercise setup
function generateRandomNumber(min, max) {
    return Math.floor(Math.random() * (max - min + 1) ) + min
}

// generateSudentEssay returns one of two versions of an assay randomly, so that
// you cannot cheat and write boolean literals based on the exercise setup
function generateSudentEssay() {
    if (generateRandomNumber(0, 10) < 5) {
        return "The expression on his face as he looked at the retro artwork demonstrates his nostalgic tendencies, especially for the times when technology was begining to pervade society."
    }
    return "The fact the retro artwork made him smile showed his deep longing for the past, especially for the time when technology began to pervade society and the outlook was still positive."
}

// generateSecretCode returns a secret code that sometimes passes validation, sometimes
// doesn't, so that you cannot cheat and write boolean literals based on the exercise setup
function generateSecretCode() {
    let code = "aj9s"
    // Add characters 5 - 9
    if (generateRandomNumber(0,10) < 7) {
        code += "Basdd"
    } else {
        code += "bmki9"
    }
    // Add characters 9 - 14
    if (generateRandomNumber(0,10) < 7) {
        code += "58yhs"
    } else {
        code += "bKols"
    }
    // Add trailing characters
    if (generateRandomNumber(0,10) < 7) {
        code += "5"
    } else {
        // Make it more than 15 characters
        code += "asd0"
    }
    return code
}
